### PR TITLE
[rtl] Various small icache bugfixes

### PR DESCRIPTION
**Pull request by tomroberts-lowrisc**
- Remove any ready -> valid dependency by allowing the skid buffer to
  accept data when the core is not ready
- Tighten-up behaviour around invalidations and cache enable/disable
- Remove xprop through output_compressed from invalid data when driving errors
- Make behaviour more consistent where speculative requests return
  different data/error conditions to existing cache hit

Signed-off-by: Tom Roberts <tomroberts@lowrisc.org>

### CODE REVIEW

File: dv/uvm/icache/dv/ibex_icache_core_agent/ibex_icache_core_protocol_checker.sv [hunk 1]
```diff
@@ -94,7 +94,7 @@ interface ibex_icache_core_protocol_checker (
           unfinished_valid |=> $stable(rdata[15:0]),
           clk, !rst_n);
   `ASSERT(HiRDataStable,
-          unfinished_valid & (rdata[1:0] == 2'b11) |=> $stable(rdata[31:16]),
+          unfinished_valid & (rdata[1:0] == 2'b11) & ~err |=> $stable(rdata[31:16]),
```
Comment (rswarbrick) [inline dv/uvm/icache/dv/ibex_icache_core_agent/ibex_icache_core_protocol_checker.sv:97]: You're right about this ("If the err signal is high, the contents of rdata don't matter"). Could we split it into a separate commit with a commit message explaining what's going on? And probably some of the other checks should have the same "& ~err". LoRDataStable, at least, I think?

File: rtl/ibex_icache.sv [hunk 2]
```diff
@@ -865,20 +876,26 @@ module ibex_icache #(
     end
   end
 
-  // The data in the skid buffer is a complete compressed instruction
-  assign skid_complete_instr = skid_valid_q & (skid_data_q[1:0] != 2'b11);
+  // The data in the skid buffer is a complete compressed instruction or if there's an error
```
Comment (rswarbrick) [inline rtl/ibex_icache.sv:879]: Maybe "The data in the skid buffer is ready if ..."?

File: rtl/ibex_icache.sv [hunk 3]
```diff
@@ -865,20 +876,26 @@ module ibex_icache #(
...
       // Once valid, the skid buffer stays valid until a compressed instruction realigns the stream
-      (skid_valid_q ? ~(ready_i & (skid_data_q[1:0] != 2'b11)) :
-      // The skid buffer becomes valid when a compressed instruction misaligns the stream
-                      ((output_addr_q[1] ^ output_compressed) & data_valid & ready_i));
+      (skid_valid_q ? ~(ready_i & ((skid_data_q[1:0] != 2'b11) | skid_err_q)) :
+      // The skid buffer becomes valid when:
+                        // - we branch to an unaligned uncompressed instruction
+                      (((output_addr_q[1] & (~output_compressed | output_err)) |
+                        // - a compressed instruction misaligns the stream
+                        (~output_addr_q[1] & output_compressed & ~output_err & ready_i)) & data_valid));
```
Comment (rswarbrick) [inline rtl/ibex_icache.sv:898]: What does the `_ | output_err` bit do here? If we branch to an unaligned uncompressed instruction and see an error, it doesn't really matter what the skid buffer contains, since the core is going to have to branch anyway, no?

File: rtl/ibex_icache.sv [hunk 4]
```diff
@@ -902,7 +919,7 @@ module ibex_icache #(
   assign output_addr_en = branch_i | (ready_i & valid_o);
 
   // Increment the address by two every time a compressed instruction is popped
-  assign addr_incr_two = output_compressed;
+  assign addr_incr_two = output_compressed & ~err_o;
```
Comment (rswarbrick) [inline rtl/ibex_icache.sv:922]: I don't think it's wrong, but why do we need this? If `err_o` is true, I would have thought our internal PC can perfectly well be 'X: it'll need resetting on the next branch anyway.

**Comment by rswarbrick** [comment]
Oh sorry, I forgot to add this. Could you also append a patch that reverts `ff5c0c5`?

File: dv/uvm/icache/dv/ibex_icache_core_agent/ibex_icache_core_protocol_checker.sv (hunk 1 above)
Comment (tomroberts-lowrisc) [reply dv/uvm/icache/dv/ibex_icache_core_agent/ibex_icache_core_protocol_checker.sv:97]: Yep will do

File: rtl/ibex_icache.sv (hunk 2 above)
Comment (tomroberts-lowrisc) [reply rtl/ibex_icache.sv:879]: Will do

File: rtl/ibex_icache.sv (hunk 3 above)
Comment (tomroberts-lowrisc) [reply rtl/ibex_icache.sv:898]: True it doesn't matter whether ``skid_valid_q`` gets set or not, but if we get an error, ``output_compressed`` will be X, which makes ``skid_valid_q`` and then various other things problematically X.

File: rtl/ibex_icache.sv (hunk 4 above)
Comment (tomroberts-lowrisc) [reply rtl/ibex_icache.sv:922]: Same as the other comment - true it'll get tidied up when the branch comes, but the X's in the address tend to propagate out problematically before then so I just mask it here.
